### PR TITLE
Add initial documentation for core APIs

### PR DESCRIPTION
**Pull request by shadaj**

### CODE REVIEW

File: core/src/lib.rs [hunk 1]
```diff
@@ -7,6 +7,8 @@ use crate::meta::types::CustomType;
 
 pub mod meta;
```
Comment (sffc) [inline core/src/lib.rs:9]: Issue: Please add module-level docs explaining what this code is for, including README and a description in Cargo.toml

File: core/src/lib.rs [hunk 2]
```diff
@@ -7,6 +7,8 @@ use crate::meta::types::CustomType;
...
 
+/// Get all custom types defined in a module as a mapping from their name to
+/// the extracted metadata.
 pub fn extract_from_mod(input: &ItemMod) -> HashMap<String, meta::types::CustomType> {
```
Comment (sffc) [inline core/src/lib.rs:12]: Nit: Please move these functions into their own module

File: core/src/meta/methods.rs [hunk 3]
```diff
@@ -2,16 +2,29 @@ use syn::*;
 
 use super::types::TypeName;
```
Comment (sffc) [inline core/src/meta/methods.rs:3]: Suggestion: `pub use` these types into *mod.rs* such that you don't need the extra level of namespace

File: core/src/ast/methods.rs [hunk 4]
```diff
@@ -2,16 +2,29 @@ use syn::*;
 
 use super::types::TypeName;
 
+/// A method declared in the `impl` associated with an FFI struct.
+/// Includes both static and non-static methods, which can be distinguished
+/// by inspecting [`self_param`].
```
Comment (Manishearth) [inline core/src/ast/methods.rs:7]: nit: I believe for the link to work it needs to be `Method::self_param`

File: core/src/ast/mod.rs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/// As part of the macro expansion and code generation process, Diplomat
+/// generates a simplified version of the Rust AST that captures special
+/// types such as opaque structs, [`Box`], and [`Result`] with utilities
+/// for handling such types.
+mod methods;
+pub use methods::*;
```
Comment (Manishearth) [inline core/src/ast/mod.rs:6]: nit: typical style is to name all items being reexported explicitly (but it's not a strong style preference)

File: core/src/ast/modules.rs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            });
+
+        Module {
+            declared_types: custom_types_by_name,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct File {
```
Comment (Manishearth) [inline core/src/ast/modules.rs:86]: thought: We'll probably eventually want this to be a crate instead (I think?) with the module map being path based (Rust paths, not file paths). Unclear; we don't have to solve multiple modules _yet_ but soon.

File: core/src/ast/types.rs [hunk 7]
```diff
@@ -6,40 +6,57 @@ use lazy_static::lazy_static;
...
+};
 
+/// A type declared inside a Diplomat-annotated module.
 #[derive(Clone, Debug)]
 pub enum CustomType {
+    /// A non-opaque struct whose fields will be visible across the FFI boundary.
     Struct(Struct),
     // TODO(shadaj): Enum
-    Opaque(String, Vec<Method>),
+    /// A struct annotated with [`diplomat::opaque` ]whose fields are not visible.
```
Comment (Manishearth) [inline core/src/ast/types.rs:20]:
```suggestion
    /// A struct annotated with [`diplomat::opaque`] whose fields are not visible.
```

File: core/src/ast/types.rs [hunk 8]
```diff
@@ -6,40 +6,57 @@ use lazy_static::lazy_static;
...
 
+/// A local type reference, such as the type of a field, parameter, or return value.
+/// Unlike [`CustomType`], which represents a type declaration, [`TypeName`]s can compose
+/// types through references and boxing, and can also capture unresolved paths.
 #[derive(Clone, Debug)]
 pub enum TypeName {
+    /// A built-in Rust scalar primitive.
     Primitive(PrimitiveType),
+    /// An unresolved path to a custom type, which can be resolved after all types
+    /// are collected with [`TypeName::resolve(env)`].
```
Comment (Manishearth) [inline core/src/ast/types.rs:50]: Does intra-doc links actually ignore things within the parentheses? I don't think it does, probably should just be `TypeName::resolve()`

**Review by Manishearth — APPROVED** [review]

**Pull request merged**